HLTrigger/Egamma/plugins/HLTEgammaDoubleXGBoostCombFilter.cc

```diff
@@ -0,0 +1,157 @@
+
+      lowMassCut_(config.getParameter<double>("lowMassCut")),
+      leadCutLowMass1_(config.getParameter<std::vector<double>>("leadCutLowMass1")),
+      subCutLowMass1_(config.getParameter<std::vector<double>>("subCutLowMass1")),
+      leadCutLowMass2_(config.getParameter<std::vector<double>>("leadCutLowMass2")),
+      subCutLowMass2_(config.getParameter<std::vector<double>>("subCutLowMass2")),
+      leadCutLowMass3_(config.getParameter<std::vector<double>>("leadCutLowMass3")),
+      subCutLowMass3_(config.getParameter<std::vector<double>>("subCutLowMass3")),
+
+      candToken_(consumes<reco::RecoEcalCandidateCollection>(config.getParameter<edm::InputTag>("candTag"))),
+      mvaToken_(consumes<reco::RecoEcalCandidateIsolationMap>(config.getParameter<edm::InputTag>("mvaPhotonTag"))) {}
+
+void HLTEgammaDoubleXGBoostCombFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+
+  desc.add<double>("highMassCut", 95.0);
+  desc.add<std::vector<double>>("leadCutHighMass1", {0.98, 0.95});
+  desc.add<std::vector<double>>("subCutHighMass1", {0.0, 0.04});
+  desc.add<std::vector<double>>("leadCutHighMass2", {0.85, 0.85});
+  desc.add<std::vector<double>>("subCutHighMass2", {0.04, 0.08});
+  desc.add<std::vector<double>>("leadCutHighMass3", {0.30, 0.50});
+  desc.add<std::vector<double>>("subCutHighMass3", {0.15, 0.20});
+
+  desc.add<double>("lowMassCut", 60.0);
+  desc.add<std::vector<double>>("leadCutLowMass1", {0.98, 0.90});
+  desc.add<std::vector<double>>("subCutLowMass1", {0.04, 0.05});
+  desc.add<std::vector<double>>("leadCutLowMass2", {0.90, 0.80});
+  desc.add<std::vector<double>>("subCutLowMass2", {0.10, 0.10});
+  desc.add<std::vector<double>>("leadCutLowMass3", {0.60, 0.60});
+  desc.add<std::vector<double>>("subCutLowMass3", {0.30, 0.30});
+
+  desc.add<edm::InputTag>("candTag", edm::InputTag("hltEgammaCandidatesUnseeded"));
+  desc.add<edm::InputTag>("mvaPhotonTag", edm::InputTag("PhotonXGBoostProducer"));
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+bool HLTEgammaDoubleXGBoostCombFilter::hltFilter(edm::Event& event,
+                                                 const edm::EventSetup& setup,
+                                                 trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  //producer collection (hltEgammaCandidates(Unseeded))
+  const auto& recCollection = event.getHandle(candToken_);
+
+  //get hold of photon MVA association map
+  const auto& mvaMap = event.getHandle(mvaToken_);
+
+  std::vector<math::XYZTLorentzVector> p4s(recCollection->size());
+  std::vector<bool> isTight(recCollection->size());
+
+  bool accept = false;
+
+  for (size_t i = 0; i < recCollection->size(); i++) {
+    edm::Ref<reco::RecoEcalCandidateCollection> refi(recCollection, i);
+    float EtaSCi = refi->eta();
+    int eta1 = (std::abs(EtaSCi) < 1.5) ? 0 : 1;
+    float mvaScorei = (*mvaMap).find(refi)->val;
```

Comment:
Just curious: what happens if the keys of `mvaMap` do not match the content of `recCollection`, and what is the intended behaviour in that case (exception, or else) ?

Naively, I would try the following.
```cpp
// get the map directly, so no need to dereference later
const auto& mvaMap = event.get(mvaToken_);
[..]
    float mvaScorei = mvaMap[refi];
```
If I read correctly [here](https://github.com/cms-sw/cmssw/blob/5a810f4abe60a5b3104aaa29a0e82aad3ddd7c63/DataFormats/Common/interface/AssociationMap.h#L189), this `operator[]` would throw a proper exception in case the reference is invalid. (It's a guess and I haven't tested it. I could be wrong.)

Reply:
Producer calculates score above the Et threshold (currently 14.25 GeV), and for others it puts -100 into the map (for that value, photon will not pass any cuts in the filter).
Producer and filter both access same collection, so the entry is always found in the map by filter.

Alternatives:
1) fill it only above the Et threshold and skip when exception is thrown and just continue looping. I didn't like this this because it could mask other problems in the configuration.
2) add the same Et threshold parameter to the filter. Then they should always stay consistent.
3) switch to collection produced by the kinematic filter. This approach is not used in other paths. Also I couldn't figure out how to make isolation map with those as candidates..

Concerning what happens when isolation is missing an entry for a candidate, I will run a test tomorrow (if I skip inserting -100 values it should throw exception or fail in some way).

Reply:
WIth the current version, if entry is missing:
```
----- Begin Fatal Exception 21-Mar-2024 11:08:57 CET-----------------------
An exception of category 'InvalidReference' occurred while
   [0] Processing  Event run: 370293 lumi: 251 event: 468138190 stream: 2
   [1] Running path 'HLT_Diphoton_MVA'
   [2] Calling method for module HLTEgammaDoubleXGBoostCombFilter/'HLTEgammaDoubleXGBoostCombFilter'
Exception Message:
can't find reference in AssociationMap at position 4
----- End Fatal Exception -------------------------------------------------
```

When using your version:
```
----- Begin Fatal Exception 21-Mar-2024 11:56:09 CET-----------------------
An exception of category 'InvalidReference' occurred while
   [0] Processing  Event run: 370293 lumi: 251 event: 468138190 stream: 0
   [1] Running path 'HLT_Diphoton_MVA'
   [2] Calling method for module HLTEgammaDoubleXGBoostCombFilter/'HLTEgammaDoubleXGBoostCombFilter'
Exception Message:
can't find reference in AssociationMap at position 3
----- End Fatal Exception -------------------------------------------------
```
If I don't induce exception, using this version also works fine.

Reply:
Okay, thanks for checking.